packages/engine/lib/error/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+//!   is inspired by `eyre`, however the problem remains the same.
+//! - [`thiserror`]? It "only" provides a macro for implementing [`Error`]. While this can be useful
+//!   for library code, handling errors can be quite tedious as the nesting gets very deep very
+//!   quickly.
+//! - [`snafu`]? Another library for generating errors using a macro. While this library is more
+//!   fully featured than `thiserror`, it is still nesting errors.
+//! - [`failure`]? [`Report`] works similar to `failure::Error`, but also `failure` is only able to
+//!   provide string-like contexts. Also `failure` uses a weird `Fail` trait instead of [`Error`].
+//!
+//! This crates does not claim to be better than the mentioned crates, it's a different approach to
+//! error handling.
+//!
+//! [`Error`]: std::error::Error
+//! [`anyhow`]: https://crates.io/crates/anyhow
+//! [`eyre`]: https://crates.io/crates/eyre
+//! [`thiserror`]: https://crates.io/crates/thiserror
+//! [`snafu`]: https://crates.io/crates/snafu
+//! [`failure`]: https://crates.io/crates/failure
+//! [`Backtrace`]: std::backtrace::Backtrace
+//! [`SpanTrace`]: tracing_error::SpanTrace
+//!
+//! # Usage
+//!
+//! [`Report`] is supposed to be used as the [`Err`] variant of a `Result`. This crates provides a
+//! [`Result`] type alias, which uses [`Report<()>`][Report] as [`Err`] variant and can be used as
+//! return type:
+//!
+//! ```
+//! # fn has_permission(_: usize, _: usize) -> bool { true }
+//! # fn get_user() -> Result<usize> { Ok(0) }
+//! # fn get_resource() -> Result<usize> { Ok(0) }
+//! use error::{ensure, Result};
+//!
+//! fn main() -> Result<()> {
+//!     let user = get_user()?;
+//!     let resource = get_resource()?;
+//!
+//!     ensure!(
+//!         has_permission(user, resource),
+//!         "Permission denied for {user} accessing {resource}"
+//!     );
+//!
+//!     # const _: &str = stringify! {
+//!     ...
+//!     # };
+//!     # Ok(())
+//! }
+//! ```
+//!
+//! A contextual message can be provided to lower level errors.
+//!
+//! ```
+//! use error::{ResultExt, Result};
+//!
+//! fn main() -> Result<()> {
+//!     # fn fake_main() -> Result<()> {
+//!     let config_path = "./path/to/config.file";
+//!     # #[cfg(all(not(miri), feature = "std"))]
+//!     # #[allow(unused_variables)]
+//!     let content = std::fs::read_to_string(config_path)
+//!         .wrap_err_lazy(|| format!("Failed to read config file {config_path:?}"))?;
+//!     # #[cfg(any(miri, not(feature = "std")))]
+//!     # Err(error::format_err!("")).wrap_err_lazy(|| format!("Failed to read config file {config_path:?}"))?;
+//!
+//!     # const _: &str = stringify! {
+//!     ...
+//!     # };
+//!     # Ok(()) }
+//!     # assert!(fake_main().is_err());
+//!     # Ok(())
+//! }
+//! ```
+//!
+//! which probably prints something like
+//!
+//! ```text
+//! Error: Failed to read config file "./path/to/config.file"
```

Comment:
Probably a good idea to split the example above so that we get two locations for the example output. Perhaps copying from below with a `read_config` function